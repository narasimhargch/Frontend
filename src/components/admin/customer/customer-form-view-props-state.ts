import { Customer } from '../../../models/customer/customer';

/* properties interface */
export interface CustomerFormViewProps {
    onSubmit: (customer: Customer) => Promise<boolean>;
}
