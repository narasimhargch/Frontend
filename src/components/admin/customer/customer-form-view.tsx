/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import { Customer } from '../../../models/customer/customer';
import { CustomerFormViewProps } from './customer-form-view-props-state';
import customerValidator from '../../../business-rules/validate-customer';

class CustomerFormView extends React.Component<CustomerFormViewProps, Customer> {
    constructor(props: CustomerFormViewProps) {
        super(props);
        this.state = { name: '', email: '' };
        this.handleEvent = this.handleEvent.bind(this);
        this.handleOnSubmit = this.handleOnSubmit.bind(this);

    }
    render(): JSX.Element {
        return (
            <div className="row">
                <div className="col-md-12">
                    <div className="row">
                        <div className="col-md-2 mb-3">
                            <label>Name</label>
                        </div>
                        <div className="col-md-4 mb-3">
                            <input 
                                type="text" 
                                name="name" 
                                className="form-control" 
                                value={this.state.name}
                                onChange={(e): void => this.handleEvent(e)}
                            />
                        </div>
                    </div>
                    <div className="row">
                        <div className="col-md-2 mb-3">
                            <label>Email</label>
                        </div>
                        <div className="col-md-4 mb-3">
                            <input 
                                type="text" 
                                name="email" 
                                className="form-control" 
                                value={this.state.email}
                                onChange={(e): void => this.handleEvent(e)}
                            />
                        </div>
                    </div>
                    <div className="row">
                        <div className="col-md-2 mb-3">
                            <label/>
                        </div>
                        <div className="col-md-4 mb-3">
                            <button 
                                type="button" 
                                onClick={(): void => this.handleOnSubmit()}
                                className="btn btn-primary"
                            >Submit
                            </button>
                        </div>
                    </div>
                </div>
            </div>
        );
    }


    private handleEvent(event: any) {
        const name = event.target.name;
        const value = event.target.value;
        this.setState({ [name]: value } as any);
    }

    private handleOnSubmit() {
        const { name, email } = this.state;
        const result = customerValidator({name: name, email: email});
        if (result) {
            const validateMsg = this.handleValidateMsg(result);

            if (validateMsg) {
                alert(validateMsg);
                return;
            } else {
                const response = this.props.onSubmit(this.state);
                if (response) {
                    this.setState({name: '', email: ''});
                }
            }
        }
    }
    
    private handleValidateMsg(result: any) {
        let validateMsg = '';

        if (result) {

            // validate customer name
            result.name && result.name.map((val: any) => {
                validateMsg += (val + '\n');
            })

            // validate email
            result.email && result.email.map((val: any) => {
                validateMsg += (val + '\n');
            })
        }
        return validateMsg;
    }
}

export default CustomerFormView;