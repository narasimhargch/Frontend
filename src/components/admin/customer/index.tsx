import {connect} from 'react-redux';
import { Dispatch } from 'redux';
import CustomerFormView from './customer-form-view';
import { Customer } from '../../../models/customer/customer';
import * as CustomerApi from '../../../services/customer-api';

const onSubmit = async (customer: Customer) => {
    const { error } = await CustomerApi.createCustomer(customer);
    return Promise.resolve(error ? false : true);
};

const mapStateToProps = () => {
    return {};
};

const mapDispatchToProps = (dispatch: Dispatch) => {
    return {
        onSubmit
    };
};

const CustomerContainer = 
    connect(mapStateToProps, mapDispatchToProps)(CustomerFormView);

export default CustomerContainer;