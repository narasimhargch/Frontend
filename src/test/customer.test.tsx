import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import CustomerFormView from '../components/admin/customer/customer-form-view';
import { Customer } from '../models/customer/customer';

test('renders learn react link', () => {
  const tree = renderer.create(<CustomerFormView onSubmit={(customer: Customer) => Promise.resolve(true)}/>).toJSON()
  expect(tree).toMatchSnapshot();
});
