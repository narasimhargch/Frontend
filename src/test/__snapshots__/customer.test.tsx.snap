// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders learn react link 1`] = `
<div
  className="row"
>
  <div
    className="col-md-12"
  >
    <div
      className="row"
    >
      <div
        className="col-md-2 mb-3"
      >
        <label>
          Name
        </label>
      </div>
      <div
        className="col-md-4 mb-3"
      >
        <input
          className="form-control"
          name="name"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-md-2 mb-3"
      >
        <label>
          Email
        </label>
      </div>
      <div
        className="col-md-4 mb-3"
      >
        <input
          className="form-control"
          name="email"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-md-2 mb-3"
      >
        <label />
      </div>
      <div
        className="col-md-4 mb-3"
      >
        <button
          className="btn btn-primary"
          onClick={[Function]}
          type="button"
        >
          Submit
        </button>
      </div>
    </div>
  </div>
</div>
`;
